File: Leetcode/2023/35_searchInsertPosition.py
class Solution(object):
    def searchInsert(self, nums, target):
        """
        :type nums: List[int]
        :type target: int
        :rtype: int
        """
        minI = 0
        maxI = len(nums) -1 
        return self.BinaSearch(nums, target, minI, maxI)
    
    def BinaSearch(self, nums, target, minI, maxI):
        minV = nums[minI]
        maxV = nums[maxI]
        midI = int (math.ceil ((minI + maxI) / 2 ) ) 

       
        print('minI: ' + str(minI) + ',maxI: ' + str(maxI) +',midI: ' + str(midI) + ' ,midV:' + str(nums[midI]))

        if target == nums[midI] :
            return midI

        elif minV >= target : 
            return minI
        elif maxV <= target :
            return maxI
        elif target < nums[midI] :
            return self.BinaSearch(nums,target, minI, midI)
        else :
            return self.BinaSearch(nums,target, midI, maxI)
        
class Solution:
    def searchInsert(self, nums: List[int], target: int) -> int:
	
        left = 0
        right = len(nums) - 1
		
		# Binary Search
		
        while left <= right:
            
			mid = (left+right)//2
			
            if nums[mid] == target:
                return mid
            elif nums[mid] < target:
                left = mid + 1
            else:
                right = mid - 1
				
        return left        

class Solution:
    def searchInsert(self, nums: List[int], target: int) -> int:
        
        left = 0
        right = len(nums) - 1
        
        # Recursive Binary Search
        def recursiveBinarySearch(nums: List[int], left, right, target: int) -> int:

            while left <= right:
                
                mid = (left+right)//2
                
                if nums[mid] == target:
                    return mid
                elif nums[mid] < target:
                    left = mid + 1
                    return recursiveBinarySearch(nums, left, right, target)
                else:
                    right = mid - 1
                    return recursiveBinarySearch(nums, left, right, target)
                
            return left
        
        return recursiveBinarySearch(nums, left, right, target)

File: Leetcode/2023/35SearchInsertPosition.java
class Solution {
//     Runtime
// 0 ms
// Beats
// 100%
// Memory
// 43.1 MB
// Beats
// 59.54%

    public int searchInsert(int[] nums, int target) {

        if(target > nums[nums.length - 1 ] ) return nums.length; 
        else if (target < nums[ 0  ] ) return 0;
        return searchBina(nums, 0, nums.length -1, target ) ;
    }

     public int searchBina (int[]nums, int left, int right, int target) {
        
        // if (target> right) return right + 1 ;
        if(left >= right ) return left ;

        int mid = (left+right) /  2 ;

        if(nums[mid] == target ) return mid ; 
        else if (nums[mid] > target ) return searchBina (nums, left, mid , target) ;
        else return searchBina(nums, mid + 1,  right , target  ) ;


    }
}